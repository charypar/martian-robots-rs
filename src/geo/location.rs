use super::orientation::Orientation;

#[derive(Eq, PartialEq, Hash, Debug)]
pub struct Point {
    pub x: i32,
    pub y: i32,
}

impl std::ops::Add for Point {
    type Output = Self;

    fn add(self, other: Self) -> Self {
        Point {
            x: self.x + other.x,
            y: self.y + other.y,
        }
    }
}

impl std::ops::Add<Orientation> for Point {
    type Output = Self;

    fn add(self, other: Orientation) -> Self {
        self + other.as_point()
    }
}
