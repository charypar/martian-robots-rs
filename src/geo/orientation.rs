use super::location::Point;

#[derive(PartialEq, Copy, Clone, Debug)]
pub enum Orientation {
    North,
    East,
    South,
    West,
}
use Orientation::{East, North, South, West};

impl Orientation {
    pub fn as_point(&self) -> Point {
        match self {
            North => Point { x: 0, y: 1 },
            East => Point { x: 1, y: 0 },
            South => Point { x: 0, y: -1 },
            West => Point { x: -1, y: 0 },
        }
    }
}

pub enum TurnDirection {
    Left,
    Right,
}
use TurnDirection::{Left, Right};

impl Orientation {
    pub fn turn(&self, rhs: TurnDirection) -> Orientation {
        match (self, rhs) {
            (orientation, Left) => match orientation {
                North => West,
                West => South,
                South => East,
                East => North,
            },
            (orientation, Right) => match orientation {
                North => East,
                East => South,
                South => West,
                West => North,
            },
        }
    }
}

#[cfg(test)]
mod tests {
    use super::Orientation::*;
    use super::TurnDirection::*;

    #[test]
    fn turns_left() {
        let pairs = [(North, West), (West, South), (South, East), (East, North)];
        for &(before, after) in &pairs {
            assert_eq!(before.turn(Left), after);
        }
    }

    #[test]
    fn turns_right() {
        let pairs = [(North, West), (West, South), (South, East), (East, North)];
        for &(before, after) in &pairs {
            assert_eq!(before.turn(Left), after);
        }
    }
}
