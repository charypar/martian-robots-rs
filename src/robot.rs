use crate::geo::location;
use crate::geo::orientation;

#[derive(PartialEq, Debug)]
pub struct Robot {
    pub position: location::Point,
    pub facing: orientation::Orientation,
}

#[derive(Clone, Copy)]
pub enum Command {
    Left,
    Right,
    Forward,
}

impl Robot {
    pub fn advance(self: Robot, command: Command) -> Robot {
        match command {
            Command::Left => Robot {
                position: self.position,
                facing: self.facing.turn(orientation::TurnDirection::Left),
            },
            Command::Right => Robot {
                position: self.position,
                facing: self.facing.turn(orientation::TurnDirection::Right),
            },
            Command::Forward => Robot {
                position: self.position + self.facing,
                facing: self.facing,
            },
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::geo::location::Point;
    use crate::geo::orientation::Orientation;

    #[test]
    fn robot_turns_left() {
        let robot = Robot {
            position: Point { x: 1, y: 2 },
            facing: Orientation::East,
        };

        let expected = Robot {
            position: Point { x: 1, y: 2 },
            facing: Orientation::North,
        };
        let actual = robot.advance(Command::Left);
        assert_eq!(actual, expected);
    }

    #[test]
    fn robot_turns_right() {
        let robot = Robot {
            position: Point { x: 1, y: 2 },
            facing: Orientation::East,
        };

        let expected = Robot {
            position: Point { x: 1, y: 2 },
            facing: Orientation::South,
        };
        let actual = robot.advance(Command::Right);
        assert_eq!(actual, expected);
    }

    #[test]

    fn robot_moves_forward() {
        let robot = Robot {
            position: Point { x: 1, y: 2 },
            facing: Orientation::East,
        };

        let expected = Robot {
            position: Point { x: 2, y: 2 },
            facing: Orientation::East,
        };
        let actual = robot.advance(Command::Forward);
        assert_eq!(actual, expected);
    }
}
