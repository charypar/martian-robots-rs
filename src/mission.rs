use std::collections::HashMap;
use std::collections::HashSet;

use crate::geo::location::Point;
use crate::geo::orientation::Orientation;
use crate::robot::Command;
use crate::robot::Robot;

#[derive(Debug, PartialEq)]
pub enum Outcome {
    Success(Robot),
    Lost(Robot),
}

pub struct Mission<I, X>
where
    I: Iterator<Item = X>,
{
    pub upper_right: Point,
    source: I,
    scents: HashMap<Point, HashSet<Orientation>>,
}

const ORIGIN: Point = Point { x: 0, y: 0 };

impl<I, SourceItem> Mission<I, SourceItem>
where
    I: Iterator<Item = SourceItem>,
{
    pub fn new(upper_right: Point, source: I) -> Mission<I, SourceItem> {
        Mission {
            upper_right: upper_right,
            source: source,
            scents: HashMap::new(),
        }
    }

    pub fn dispatch(&mut self, robot: Robot, commands: &[Command]) -> Outcome {
        let outcome = commands.iter().try_fold(robot, |r, c| {
            let robot = r.advance(*c);

            if (ORIGIN.x..=self.upper_right.x).contains(&robot.position.x)
                && (ORIGIN.y..=self.upper_right.y).contains(&robot.position.y)
            {
                // moved robot is still on the grid, commit
                return Ok(robot);
            }

            // moved robot would be off the grid...
            match self.scents.get(&r.position) {
                // ...but previous robot has left a scent, so we'll ignore the move
                Some(scent) if scent.contains(&robot.facing) => return Ok(r),
                // ...and it's lost, but not before leaving a scent in its wake
                _ => {
                    self.scents
                        .entry(r.position)
                        .or_insert(HashSet::new())
                        .insert(r.facing);
                    Err(r)
                }
            }
        });

        match outcome {
            Ok(robot) => Outcome::Success(robot),
            Err(robot) => Outcome::Lost(robot),
        }
    }
}

// Running a mission with a reliable source
impl<I> std::iter::Iterator for Mission<I, (Robot, Vec<Command>)>
where
    I: Iterator<Item = (Robot, Vec<Command>)>,
{
    type Item = Outcome;

    fn next(&mut self) -> Option<Self::Item> {
        self.source
            .next()
            .map(|(robot, commands)| self.dispatch(robot, commands.as_ref()))
    }
}

// Running a mission with a unreliable source
impl<I> std::iter::Iterator for Mission<I, Result<(Robot, Vec<Command>), String>>
where
    I: Iterator<Item = Result<(Robot, Vec<Command>), String>>,
{
    type Item = Result<Outcome, String>;

    fn next(&mut self) -> Option<Self::Item> {
        self.source
            .next()
            .map(|item| item.map(|(robot, commands)| self.dispatch(robot, commands.as_ref())))
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::geo::location::Point;
    use crate::geo::orientation::Orientation::{East, North, South, West};
    use crate::robot::Command::{Forward as F, Left as L, Right as R};
    use crate::robot::Robot;

    #[test]
    fn simple_robot() {
        let mut mission: Mission<_, (Robot, Vec<Command>)> =
            Mission::new(Point { x: 5, y: 3 }, Vec::new().into_iter());
        let robot = Robot {
            position: Point { x: 1, y: 1 },
            facing: East,
        };

        let expected = Outcome::Success(Robot {
            position: Point { x: 1, y: 1 },
            facing: East,
        });
        let actual = mission.dispatch(robot, &[R, F, R, F, R, F, R, F]);

        assert_eq!(actual, expected);
    }

    #[test]
    fn robot_is_lost() {
        let mut mission: Mission<_, (Robot, Vec<Command>)> =
            Mission::new(Point { x: 5, y: 3 }, Vec::new().into_iter());
        let robot = Robot {
            position: Point { x: 3, y: 2 },
            facing: North,
        };

        let expected = Outcome::Lost(Robot {
            position: Point { x: 3, y: 3 },
            facing: North,
        });
        let actual = mission.dispatch(robot, &[F, R, R, F, L, L, F, F, R, R, F, L, L]);

        assert_eq!(actual, expected);
    }

    #[test]
    fn robots_are_clever() {
        let mut mission: Mission<_, (Robot, Vec<Command>)> =
            Mission::new(Point { x: 5, y: 3 }, Vec::new().into_iter());
        let robot = Robot {
            position: Point { x: 3, y: 2 },
            facing: North,
        };

        let expected = Outcome::Lost(Robot {
            position: Point { x: 3, y: 3 },
            facing: North,
        });
        let actual = mission.dispatch(robot, &[F, R, R, F, L, L, F, F, R, R, F, L, L]);

        assert_eq!(actual, expected);

        let robot = Robot {
            position: Point { x: 0, y: 3 },
            facing: West,
        };

        let expected = Outcome::Success(Robot {
            position: Point { x: 2, y: 3 },
            facing: South,
        });
        let actual = mission.dispatch(robot, &[L, L, F, F, F, L, F, L, F, L]);

        assert_eq!(actual, expected);
    }
}
