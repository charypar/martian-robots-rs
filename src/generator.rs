use rand::distributions::Standard;
use rand::prelude::*;
use rand::rngs::SmallRng;

use crate::geo::location::Point;
use crate::geo::orientation::Orientation;
use crate::mission::Mission;
use crate::robot::{Command, Robot};

struct Generator {
  pub mission: Mission,
  prng: SmallRng, // a pseudo random number generator
}

impl Generator {
  pub fn new(seed: u64) -> Generator {
    let mut prng = SmallRng::seed_from_u64(seed);
    let mission = Mission::new(Point {
      x: prng.gen_range(1, 51),
      y: prng.gen_range(1, 51),
    });

    Generator { mission, prng }
  }
}

impl Iterator for Generator {
  type Item = (Robot, Vec<Command>);

  fn next(&mut self) -> Option<Self::Item> {
    let rng = &mut self.prng;

    let robot = Robot {
      position: Point {
        x: rng.gen_range(0, self.mission.upper_right.x),
        y: rng.gen_range(0, self.mission.upper_right.y),
      },
      facing: rng.gen(),
    };

    let ncmds = rng.gen_range(1, 100);
    let commands = rng.sample_iter(Standard).take(ncmds).collect();

    Some((robot, commands))
  }
}

// Randomly generated custom types

impl Distribution<Orientation> for Standard {
  fn sample<R: Rng + ?Sized>(&self, rng: &mut R) -> Orientation {
    match rng.gen_range(0, 4) {
      0 => Orientation::North,
      1 => Orientation::West,
      2 => Orientation::South,
      _ => Orientation::East,
    }
  }
}

impl Distribution<Command> for Standard {
  fn sample<R: Rng + ?Sized>(&self, rng: &mut R) -> Command {
    match rng.gen_range(0, 3) {
      0 => Command::Left,
      1 => Command::Right,
      _ => Command::Forward,
    }
  }
}

#[cfg(test)]
mod tests {
  use super::*;
  use crate::geo::orientation::Orientation::{East as E, North as N, South as S, West as W};
  use crate::robot::Command::{Forward as F, Left as L, Right as R};

  #[test]
  fn creates_mission() {
    let generator = Generator::new(12345);
    let expected = Point { x: 17, y: 43 };

    assert_eq!(generator.mission.upper_right, expected);
  }

  #[test]
  fn generates_three_robots() {
    let generator = Generator::new(12345);
    let expected = vec![
      (
        Robot {
          position: Point { x: 16, y: 39 },
          facing: S,
        },
        vec![
          R, L, R, L, R, L, R, L, F, F, R, L, R, R, F, R, L, F, L, R, L, F, R, F, R, R, F, L, F, L,
          L, F, L, F, F, L, F, R, L, L, F, R, R, R, L, F, R, L, R, R, R, L, F, R, L, L, F, R,
        ],
      ),
      (
        Robot {
          position: Point { x: 12, y: 32 },
          facing: S,
        },
        vec![
          L, R, F, F, F, R, L, R, L, L, R, L, R, L, F, F, R, F, L, F, L, F, L, R, R, F, F, F, R, F,
          F, F, L, F, R, L, F, F, L, L, F, R, L, R, F, F, F, F, R, R, R, R, R, F, F, L, R, R, L, R,
          F, R, L, R, F, R, F, L, R, L, L, R, L, R,
        ],
      ),
      (
        Robot {
          position: Point { x: 7, y: 42 },
          facing: N,
        },
        vec![
          L, L, R, L, L, R, R, R, R, L, F, F, F, R, L, R, L, F, L, F, L, R, F, R, L, R, L, L, R, F,
          R, R, R, R,
        ],
      ),
    ];

    assert_eq!(generator.take(3).collect::<Vec<_>>(), expected);
  }
}
