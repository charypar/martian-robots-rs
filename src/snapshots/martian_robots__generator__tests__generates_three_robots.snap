---
source: src/generator.rs
expression: "generator.take(3).collect::<Vec<_>>()"
---
[
    (
        Robot {
            position: Point {
                x: 16,
                y: 39,
            },
            facing: South,
        },
        [
            Right,
            Left,
            Right,
            Left,
            Right,
            Left,
            Right,
            Left,
            Forward,
            Forward,
            Right,
            Left,
            Right,
            Right,
            Forward,
            Right,
            Left,
            Forward,
            Left,
            Right,
            Left,
            Forward,
            Right,
            Forward,
            Right,
            Right,
            Forward,
            Left,
            Forward,
            Left,
            Left,
            Forward,
            Left,
            Forward,
            Forward,
            Left,
            Forward,
            Right,
            Left,
            Left,
            Forward,
            Right,
            Right,
            Right,
            Left,
            Forward,
            Right,
            Left,
            Right,
            Right,
            Right,
            Left,
            Forward,
            Right,
            Left,
            Left,
            Forward,
            Right,
        ],
    ),
    (
        Robot {
            position: Point {
                x: 12,
                y: 32,
            },
            facing: South,
        },
        [
            Left,
            Right,
            Forward,
            Forward,
            Forward,
            Right,
            Left,
            Right,
            Left,
            Left,
            Right,
            Left,
            Right,
            Left,
            Forward,
            Forward,
            Right,
            Forward,
            Left,
            Forward,
            Left,
            Forward,
            Left,
            Right,
            Right,
            Forward,
            Forward,
            Forward,
            Right,
            Forward,
            Forward,
            Forward,
            Left,
            Forward,
            Right,
            Left,
            Forward,
            Forward,
            Left,
            Left,
            Forward,
            Right,
            Left,
            Right,
            Forward,
            Forward,
            Forward,
            Forward,
            Right,
            Right,
            Right,
            Right,
            Right,
            Forward,
            Forward,
            Left,
            Right,
            Right,
            Left,
            Right,
            Forward,
            Right,
            Left,
            Right,
            Forward,
            Right,
            Forward,
            Left,
            Right,
            Left,
            Left,
            Right,
            Left,
            Right,
        ],
    ),
    (
        Robot {
            position: Point {
                x: 7,
                y: 42,
            },
            facing: North,
        },
        [
            Left,
            Left,
            Right,
            Left,
            Left,
            Right,
            Right,
            Right,
            Right,
            Left,
            Forward,
            Forward,
            Forward,
            Right,
            Left,
            Right,
            Left,
            Forward,
            Left,
            Forward,
            Left,
            Right,
            Forward,
            Right,
            Left,
            Right,
            Left,
            Left,
            Right,
            Forward,
            Right,
            Right,
            Right,
            Right,
        ],
    ),
]
